firefox_ios/views/client_counts.view.lkml

```diff
@@ -0,0 +1,60 @@
+include: "//looker-hub/firefox_ios/views/baseline_clients_daily_table.view.lkml"
+
+view: client_counts {
+  extends: [baseline_clients_daily_table]
+
+  dimension_group: since_first_seen {
+    type: duration
+    description: "Amount of time that has passed since the client was first seen."
+    sql_start: ${TABLE}.first_seen_date ;;
+    sql_end: ${TABLE}.submission_date ;;
+    intervals: [
+      day, week, month, year
+    ]
+  }
+
+  dimension: have_completed_period {
+    type: yesno
+    description: "
+      Only for use with cohort analysis.
+      Filter on true to remove the tail of incomplete data from cohorts.
+      Indicates whether the cohort for this row have all had a chance to complete this interval.
+      For example, new clients from yesterday have not all had a chance to send a ping for today."
+    sql:
+      DATE_ADD(
+        {% if client_counts.first_seen_date._is_selected %}
+          DATE_ADD(DATE(${client_counts.first_seen_date}), INTERVAL 1 DAY)
+        {% elsif client_counts.first_seen_week._is_selected %}
+          DATE_ADD(DATE(${client_counts.first_seen_week}), INTERVAL 1 WEEK)
+        {% elsif client_counts.first_seen_month._is_selected %}
+          DATE_ADD(DATE(${client_counts.first_seen_month}), INTERVAL 1 MONTH)
+        {% elsif client_counts.first_seen_year._is_selected %}
+          DATE_ADD(DATE(${client_counts.first_seen_year}), INTERVAL 1 YEAR)
+        {% endif %}
+        ,
+        {% if client_counts.days_since_first_seen._is_selected %}
+          INTERVAL ${client_counts.days_since_first_seen} DAY
+        {% elsif client_counts.weeks_since_first_seen._is_selected %}
+          INTERVAL ${client_counts.weeks_since_first_seen} WEEK
+        {% elsif client_counts.months_since_first_seen._is_selected %}
+          INTERVAL ${client_counts.months_since_first_seen} MONTH
+        {% elsif client_counts.years_since_first_seen._is_selected %}
+          INTERVAL ${client_counts.months_since_first_seen} YEAR
+        {% endif %}
+      ) < current_date
```

Comment:
I know this is a real beast. But let's start with an example.

If we're cohorting on weeks, and we want to see, each week from cohort creation, what fraction of the cohort is still active. Then we would not want to include data for week 1 from last week's cohort, since this week hasn't finished - they would look artificially deflated in retention.

For that case, we have:
```
DATE_ADD(
  -- The following gives the first date after the end of this cohort's first_week
  DATE_ADD(DATE(${client_counts.first_seen_week}), INTERVAL 1 WEEK),
  -- Then we add the current period, e.g. are we looking two weeks in the future?
  INTERVAL ${client_counts.weeks_since_first_seen} WEEK
)
-- Now we check to see if we have the data for that period
< current_date
```

Reply:
This would actually be a nice time to use that `last_etl_date` UDF, since there will be a few hours this is not correct, while ETL finishes before the day starts.